apps/omg/test/omg/conformance/merkle_proof_property_test.exs

```diff
@@ -0,0 +1,98 @@
+  implementations (currently `elixir-omg` and `plasma-contracts`, Elixir and Solidity)
+  """
+
+  alias OMG.Merkle
+  alias Support.Conformance.MerkleProofContext
+  alias Support.SnapshotContracts
+
+  import Support.Conformance.MerkleProofs, only: [solidity_proof_valid: 5]
+
+  use PropCheck
+  use ExUnit.Case, async: false
+
+  @moduletag :property
+  @moduletag timeout: 450_000
+
+  setup_all do
+    {:ok, exit_fn} = Support.DevNode.start()
+
+    contracts = SnapshotContracts.parse_contracts()
+    merkle_wrapper_address_hex = contracts["CONTRACT_ADDRESS_MERKLE_WRAPPER"]
+
+    on_exit(exit_fn)
+
+    [contract: OMG.Eth.Encoding.from_hex(merkle_wrapper_address_hex)]
+  end
+
+  property "any root hash and proof created by the Elixir implementation validates in the contract, for all leaves",
+           [250, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall leaves <- list(binary()) do
+      root_hash = Merkle.hash(leaves)
+
+      leaves
+      |> Enum.with_index()
+      |> Enum.all?(fn {leaf, txindex} ->
+        proof = Merkle.create_tx_proof(leaves, txindex)
+        solidity_proof_valid(leaf, txindex, root_hash, proof, contract)
+      end)
+    end
+  end
+
+  property "no proof can prove a mutated leaf",
+           [2500, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall proof <- MerkleProofContext.correct() do
+      forall mutated <- MerkleProofContext.mutated_leaf(proof) do
+        not solidity_proof_valid(mutated.leaf, mutated.txindex, mutated.root_hash, mutated.proof, contract)
+      end
+    end
+  end
+
+  property "no proof can prove at different index",
+           [2500, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall proof <- MerkleProofContext.correct() do
+      forall mutated <- MerkleProofContext.mutated_txindex(proof) do
+        not solidity_proof_valid(mutated.leaf, mutated.txindex, mutated.root_hash, mutated.proof, contract)
+      end
+    end
+  end
+
+  property "no mutated proof bytes can prove anything that the original proved",
+           [2500, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall proof <- MerkleProofContext.correct() do
+      forall mutated <- MerkleProofContext.mutated_proof(proof) do
+        not solidity_proof_valid(mutated.leaf, mutated.txindex, mutated.root_hash, mutated.proof, contract)
+      end
+    end
+  end
+
+  property "no proof can prove a different leaf/txindex if proof bytes mutated",
+           [2500, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall proof <- MerkleProofContext.correct() do
+      forall mutated <- MerkleProofContext.mutated_to_prove_sth_else(proof) do
+        not solidity_proof_valid(mutated.leaf, mutated.txindex, mutated.root_hash, mutated.proof, contract)
+      end
+    end
+  end
+end
```

Comment:
I've tried a following test that has demonstrated two things:
1. mutations do not include originals
2. tree generator allows duplicates in leaves (see also proposed change for generator)

```
+  property "valid proof can't prove at different index",
+           [2500, :verbose, max_size: 256, constraint_tries: 100_000],
+           %{contract: contract} do
+    forall proof <- MerkleProofContext.correct() do
+      forall index <- pos_integer() do
+        not solidity_proof_valid(proof.leaf, index, proof.root_hash, proof.proof, contract) or
+          index == proof.txindex
+      end
+    end
+  end
```

Reply:
both 1. and 2. are intentional, not sure what are you suggesting here?

Reply:
since there's a bunch of areas how the prop tests and generators can be expanded and improved, I'll take the liberty of including links to your suggestions in the comments, to revisit later. Thx! :pray: